planner/core/stats.go

```diff
@@ -180,7 +181,9 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo, selfSchema *
 		}
 	}
 	// Consider the IndexMergePath. Now, we just generate `IndexMergePath` in DNF case.
-	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 && ds.ctx.GetSessionVars().GetEnableIndexMerge() {
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 &&
+		(ds.ctx.GetSessionVars().GetEnableIndexMerge() ||
+			(ds.indexMergeHints != nil) && !ds.ctx.GetSessionVars().StmtCtx.NoIndexMergeHint) {
```

Comment:
I suggest that this long `if` to be split into two parts: one is can we use index merge, which is decided by hints or session vars; other one is do we need to use index merge, which is decided by conds. It would be easier for read.
And what if we have a no index merge hint, and session variable `enable index merge` is on, in this case I think we should not use index merge, but currently it will?